libs/wire-api/src/Wire/API/OAuth.hs

```diff
@@ -0,0 +1,596 @@
+
+instance ToSchema OAuthClientPlainTextSecret where
+  schema = (toText . unOAuthClientPlainTextSecret) .= parsedText "OAuthClientPlainTextSecret" (fmap OAuthClientPlainTextSecret . validateBase16)
+
+instance FromHttpApiData OAuthClientPlainTextSecret where
+  parseQueryParam = bimap cs OAuthClientPlainTextSecret . validateBase16 . cs
+
+instance ToHttpApiData OAuthClientPlainTextSecret where
+  toQueryParam = toText . unOAuthClientPlainTextSecret
+
+data OAuthClientCredentials = OAuthClientCredentials
+  { occClientId :: OAuthClientId,
+    occClientSecret :: OAuthClientPlainTextSecret
+  }
+  deriving (Eq, Show, Generic)
+  deriving (A.ToJSON, A.FromJSON, S.ToSchema) via (Schema OAuthClientCredentials)
+
+instance ToSchema OAuthClientCredentials where
+  schema =
+    object "OAuthClientCredentials" $
+      OAuthClientCredentials
+        <$> occClientId .= fieldWithDocModifier "client_id" clientIdDescription schema
+        <*> occClientSecret .= fieldWithDocModifier "client_secret" clientSecretDescription schema
+    where
+      clientIdDescription = description ?~ "The ID of the application."
+      clientSecretDescription = description ?~ "The secret of the application."
+
+data OAuthClient = OAuthClient
+  { ocId :: OAuthClientId,
+    ocName :: OAuthApplicationName,
+    ocRedirectUrl :: RedirectUrl
+  }
+  deriving (Eq, Show, Generic)
+  deriving (A.ToJSON, A.FromJSON, S.ToSchema) via (Schema OAuthClient)
+
+instance ToSchema OAuthClient where
+  schema =
+    object "OAuthClient" $
+      OAuthClient
+        <$> ocId .= field "client_id" schema
+        <*> ocName .= field "application_name" schema
+        <*> ocRedirectUrl .= field "redirect_url" schema
+
+data OAuthResponseType = OAuthResponseTypeCode
+  deriving (Eq, Show, Generic)
+  deriving (A.ToJSON, A.FromJSON, S.ToSchema) via (Schema OAuthResponseType)
+
+instance ToSchema OAuthResponseType where
+  schema :: ValueSchema NamedSwaggerDoc OAuthResponseType
+  schema =
+    enum @Text "OAuthResponseType" $
+      mconcat
+        [ element "code" OAuthResponseTypeCode
+        ]
+
+-- | The OAuth scopes that are supported by the backend.
+-- This type is a bit redundant and unfortunately has to be kept in sync
+-- with the supported scopes defined in the nginx configs.
+-- However, having this typed makes it easier to handle scopes in the backend,
+-- and e.g. provide more meaningful error messages when the scope is invalid.
+data OAuthScope
+  = ReadFeatureConfigs
+  | ReadSelf
+  | WriteConversation
+  | WriteConversationCode
+  deriving (Eq, Show, Generic, Ord)
+  deriving (Arbitrary) via (GenericUniform OAuthScope)
+
+instance ToByteString OAuthScope where
+  builder = \case
+    WriteConversation -> "write:conversations"
+    WriteConversationCode -> "write:conversations_code"
+    ReadSelf -> "read:self"
+    ReadFeatureConfigs -> "read:feature_configs"
+
+instance FromByteString OAuthScope where
+  parser = do
+    s <- parser
+    case s & T.toLower of
```

Comment:
```suggestion
    case T.toLower s of
```

Reply:
Hehe, I like it with the `&` because it reads better (left to right), but I don't mind. :)